README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# clustena

<!-- badges: start -->

[![R-CMD-check](https://github.com/lkhenayfis/hidro-clustena/workflows/R-CMD-check/badge.svg)](https://github.com/lkhenayfis/hidro-clustena/actions)
[![test-coverage](https://github.com/lkhenayfis/hidro-clustena/workflows/test-coverage/badge.svg)](https://github.com/lkhenayfis/hidro-clustena/actions)
<!-- badges: end -->

Este pacote prove funcoes para selecao de cenarios de Energia Natural
Afluente, independentemente da metodologia utilizada para gera-los. Alem
dos metodos para selecao de cenarios sao fornecidas ferramentas para
visualizacao da escolha.

## Instalacao

Este pacote ainda nao se encontra disponibilizado no CRAN, de modo que
deve ser instalado diretamente a partir do repositorio utilizando:

``` r
# Caso a biblioteca remotes nao esteja instalada, execute install.packages("remotes") primeiro
remotes::install_github("lkhenayfis/hidro-clustena", build_vignettes = FALSE)
```

## Exemplo de uso

Abaixo esta um trecho de codigo exemplificando de forma simplificada o
uso das funcionalidades contidas neste pacote. Mais detalhes acerca das
funcoes se encontram em suas respectivas paginas de ajuda.

``` r
library(clustena)

# selecao de cenarios por PCA + Kmeans (com dado exemplo do pacote)
cenarios <- cenariosdummy["A1", "SIN", , -c(1:5)]
selected <- selecporcluster(cenarios, 3)
plot(cenarios, cenarios[, , selected[[1]]])
```

<img src="man/figures/README-leitura-1.png" width="100%" />


NEWS.md
# clustena 1.3.2

## New features

* Inclui diretorios `conf` e `scripts` em `inst`, contendo script para rodada operacional da selecao
  de cenarios e arquivo de configuracao default do mesmo, respectivamente
* `jsonlite` foi adicionado aos Suggests para realizar a leitura dos arquivos de configuracao
* Funcoes de compactacao agora tratam todas as bacias inclusas no dado de forma multivariada:
  * `acumulaena` faz a soma acumulada de cada bacia em cada cenario, retornando uma dimensao 
    compactada igual a `Nbacias * Nquebras`
  * `PCAena` concatena os trajetos de cada bacia em uma observacao `tamanho_cen * Nbacias` e usa
    isso como dado para compactar
* `extracdims` agora considera a possibilidade de multiplas bacias no dado. A compactacao por PCA
  naturalmente ja devolvia os cenarios no espaco reduzido considerando todas as bacias originais,
  mas a acumulacao ainda devolvia um espaco compactado por bacia. Isto causava um erro
* Introduz a funcao `clusthierarq` para clusterizacao na selecao/extracao de cenarios por metodos
  aglomerativos
* Introduz a funcao `clustkmedoids` para clusterizacao na selecao/extracao de cenarios por 
  k-medoides

### Minor

* Introduz funcao `as.cenariosena` para conversao de `data.table`s ja lidos em `cenariosena`
* Melhora documentacao de `PCAena` e `acumulaena`
* Adiciona exemplos nas documentacoes do pacote

## Bug fixes

* Corrige um erro em `selecporcluster` que ocorria quando clusters tinham apenas um elemento. O 
  calculo do elemento mais proximo dava erro em funcao da simplificacao de matriz para vetores

### Minor

* Corrige documentacao dos metodos `getclustclass` e `getclustmeans` para `Mclust` e `kmeans`

# clustena 1.0

Este pacote prove funcoes para selecao de cenarios de Energia Natural Afluente, independentemente da
metodologia utilizada para gera-los. Alem dos metodos para selecao de cenarios sao fornecidas
ferramentas para visualizacao da escolha.


tests/testthat/test-leitura.r
test_that("Testes de leitura e classe cenariosena", {

    # usando arquivo interno
    arq  <- system.file("extdata/cenarios.csv", package = "clustena")
    cens <- learqcenarios(arq)

    # Testes da classe cenariosena

    expect_equal(class(cens), "cenariosena")

    expect_equal(class(cens[[1]]), c("data.table", "data.frame"))
    expect_equal(colnames(cens[[1]]), c("anoref", "bacia", "cenario", "data", "ena"))

    expect_equal(attr(cens, "bacias"), c("N", "NE", "SE", "SIN", "SUL"))
    expect_equal(attr(cens, "datas"), seq(as.Date("2022-09-01"), as.Date("2023-12-01"), "month"))
    expect_equal(attr(cens, "anos"), paste0("A", seq(4)))
    expect_equal(attr(cens, "ncen"), 100)

    # Testes de subset

    cc <- cens[c("A1", "A4"), c("SUL", "NE"), 40:70, c("2022-11-01", "2023-03-01", "2023-07-01")]

    expect_equal(attr(cc, "bacias"), c("NE", "SUL"))
    expect_equal(attr(cc, "datas"), as.Date(c("2022-11-01", "2023-03-01", "2023-07-01")))
    expect_equal(attr(cc, "anos"), c("A1", "A4"))
    expect_equal(attr(cc, "ncen"), 31)

    gg <- plot(cc, print = FALSE)
    expect_equal(class(gg), c("gg", "ggplot"))

    gg <- plot(cc, cc["A1", "SUL", c(55, 60, 65)], cc["A4", "NE", c(50, 55, 60)], print = FALSE)
    expect_equal(class(gg), c("gg", "ggplot"))
})

tests/testthat/test-selecao.r
test_that("Testes de selecao de cenarios", {

    set.seed(1234)

    cens <- cenariosdummy["A1", "SIN"]

    sel <- selecporcluster(cens, 5)

    expect_equal(length(sel), 3)

    expect_equal(class(sel[[1]]), "integer")
    expect_equal(length(sel[[1]]), 5)

    expect_equal(class(sel[[2]]), "compactcen")
    expect_equal(class(sel[[3]]), "kmeans")

    sel <- selecporcluster(cens, 5, compact_fun = acumulaena, compact_args = list(quebras = 3),
        clust_fun = clustEM, verbose = FALSE)

    expect_equal(length(sel), 3)

    expect_equal(class(sel[[1]]), "integer")
    expect_equal(length(sel[[1]]), 5)

    expect_equal(class(sel[[2]]), "compactcen")
    expect_equal(class(sel[[3]]), "Mclust")
})

tests/testthat/test-compactacao.r
test_that("Testes das funcoes de compactacao", {

    cens <- cenariosdummy["A1", "SIN"]

    # PCA

    compac <- PCAena(cens, vartot = 1)

    expect_equal(colnames(compac$compact), c("anoref", "bacia", "cenario", "ind", "ena"))
    expect_equal(class(compac), "compactcen")
    expect_equal(attr(compac, "metodo"), "PCAena")
    expect_true(attr(compac, "teminv"))

    reverse <- unname(attr(compac, "invfunc")(compac$compact[cenario == 1, ena]))[1, ]
    expect_snapshot_value(reverse, style = "deparse")

    expect_error(plot(compac, print = FALSE))

    compac <- PCAena(cens)

    gg <- plot(compac, print = FALSE)
    expect_equal(class(gg), c("plotly", "htmlwidget"))

    compac <- PCAena(cens, .75)

    gg <- plot(compac, print = FALSE)
    expect_equal(class(gg), c("gg", "ggplot"))

    # ACUMULADA

    compac <- acumulaena(cens, quebras = 5L)

    expect_equal(colnames(compac$compact), c("anoref", "bacia", "cenario", "ind", "ena"))
    expect_equal(class(compac), "compactcen")
    expect_equal(attr(compac, "metodo"), "acumulaena")
    expect_true(!attr(compac, "teminv"))

    expect_error(plot(compac, print = FALSE))

    compac <- acumulaena(cens, 3L)

    gg <- plot(compac, print = FALSE)
    expect_equal(class(gg), c("plotly", "htmlwidget"))

    compac <- acumulaena(cens, 2L)

    gg <- plot(compac, print = FALSE)
    expect_equal(class(gg), c("gg", "ggplot"))
})

R/selecao.r
######################### FUNCOES PARA CLUSTERIZAR E/OU SELECIONAR CENARIOS ########################

#' Selecao De Cenarios Por Quantil
#' 
#' Seleciona cenarios representativos buscando aqueles mais proximos de quantis especificados
#' 
#' A selecao por quantis requer que cada cenario seja reduzido a apenas uma dimensao (de modo que 
#' seja possivel calcular quantis unicos). A compactacao pode ser feita por qualquer funcao que
#' retorne objetos da classe \code{compactcen}, informada atraves do argumento \code{compact_fun}. A
#' funcao sera chamada com \code{cenarios} como primeiro arumento; \code{compact_args} permite 
#' informar demais parametros a serem utilizados por \code{compact_fun}, na forma de uma lista 
#' nomeada.
#' 
#' O argumento cenarios deve ser um objeto da classe \code{cenariosena} contendo apenas um ano de 
#' referencia e bacia. \code{quantis} deve ser um vetor indicando quais quantis estao sendo 
#' buscados. Por construcao e impossivel extrair exatamente estes quantis, de modo que serao 
#' retornados os indices dos cenarios associados aos quantis mais proximos. 
#' 
#' @param cenarios objeto da classe \code{cenariosena} do qual selecionar cenarios representativos
#'     contendo apenas uma bacia e ano de referencia
#' @param quantis vetor de quantis para selecionar
#' @param compact_fun funcao para compactacao dos cenarios. Ver \code{\link{compact_funs}}
#' @param compact_args lista nomeada informando argumentos (alem do primeiro) a \code{compact_fun}
#' 
#' @return lista contendo vetor de inteiros indicando indices dos cenarios selecionados e resultado 
#'     da compactacao
#' 
#' @examples 
#' 
#' # usando o dado exemplo do pacote
#' 
#' # a selecao deve ser usada para cada ano de referencia individualmente
#' cens <- cenariosdummy["A1"]
#' 
#' # execucao simples
#' selec <- selecporquantil(cens)
#' 
#' # pegando quantis c(.33, .5, .67)
#' selec <- selecporquantil(cens, c(.33, .5, .67))
#' 
#' @seealso \code{\link{plot.cenariosena}} para visualizacao dos cenarios selecionados
#' 
#' @export

selecporquantil <- function(cenarios, quantis = c(.25, .5, .75), compact_fun = acumulaena,
    compact_args = list()) {

    compact_call <- c(list(compact_fun, cenarios), compact_args)
    compact <- eval(as.call(compact_call))

    dat <- compact$compact
    dat <- dcast(dat, cenario ~ ind, value.var = "ena")[, -1]

    if(ncol(dat) > 1) stop("A compactacao para selecao por quantil deve ser em apenas uma dimensao")

    dat_quant <- quantile(dat[[1]], quantis)
    maisprox  <- unname(sapply(dat_quant, function(dq) which.min((dat[[1]] - dq)^2)))

    out <- list(maisprox, compact)

    return(out)
}

#' Selecao Ou Extracao De Cenarios Por Clusterizacao
#' 
#' Seleciona ou extrai cenarios baseado nos centroides da clusterizacao de cenarios
#' 
#' O procedimento desta funcao e aplicar a reducao de dimensionalidade em \code{cenarios} e 
#' clusterizar as variaveis no espaco reduzido em \code{nc} clusters.
#' 
#' A compactacao pode ser feita por qualquer funcao que retorne objetos da classe \code{compactcen},
#' informada atraves do argumento \code{compact_fun}. A funcao sera chamada com \code{cenarios} como
#' primeiro arumento; \code{compact_args} permite informar demais parametros a serem utilizados por
#' \code{compact_fun}, na forma de uma lista nomeada.
#' 
#' A clusterizacao em si e realizada por uma chamada a funcao passada pelo argumento 
#' \code{clust_fun}, uma funcao que deve receber como primeiro argumento um objeto \code{cenarios} e 
#' como segundo argumento o numero de clusters desejado \code{nc}, e retornar um objeto com os 
#' metodos \code{getclustclass} e \code{getclustmeans}, para retornar as classificacoes de cada 
#' cenario e as medias dos centroides. A funcao passada sera usada numa chamada da forma 
#' \code{clust_fun(cenarios, nc, ...)}. \code{compact_fun} e \code{compact_args}.
#' 
#' Por fim, \code{transforma} controla o que sera retornado. Se \code{FALSE}, a funcao entrega 
#' apenas os indices dos cenarios mais proximos, por distancia euclidiana, dos centroides de cada
#' cluster. Se \code{TRUE}, o centroide de cada cluster (no espaco reduzido) sera transformado de
#' volta para o espaco original. Isso so e possivel caso a compactacao utilizada permita este tipo
#' de operacao: por exemplo, a reducao de dimensionalidade por acumulo de enas parciais e uma 
#' transformacao que nao permite retornar para a escala original. Outras como PCA e autoencoder sim.
#' 
#' @param cenarios objeto da classe \code{cenariosena} do qual selecionar cenarios representativos
#'     contendo apenas uma bacia e ano de referencia
#' @param nc numero de clusters
#' @param clust_fun funcao para clusterizacao. Veja \code{\link{clust_funs}} para mais detalhes
#' @param ... demais parametros passados a \code{clust_fun}
#' @param transforma booleano indicando se devem ser retornados indices dos cenarios mais proximos
#'     do centroide de cada cluster (\code{FALSE}) ou se o cetroide deve ser transformado de volta
#'     para a escala original (\code{TRUE}). Ver Detalhes
#' @param compact_fun funcao para compactacao dos cenarios. Ver \code{\link{compact_funs}}
#' @param compact_args lista nomeada informando argumentos (alem do primeiro) a \code{compact_fun}
#' 
#' @return lista contendo vetor de inteiros indicando indices dos cenarios selecionados, resultado 
#'     da compactacao e resultado da clusterizacao
#' 
#' @examples
#' 
#' # usando o dado exemplo do pacote
#' 
#' # a selecao deve ser usada para cada ano de referencia individualmente, para uma unica regiao
#' # ou multiplas de forma multivariada
#' # Usando somente o SIN completo
#' cens <- cenariosdummy["A1", "SIN"]
#' 
#' # execucao simples pegando 3 clusters
#' selec <- selecporcluster(cens, 3)
#' 
#' # usando diferentes funcoes de compactacao
#' selec <- selecporcluster(cens, 3, compact_fun = acumulaena, compact_args = list(quebras = 3))
#' selec <- selecporcluster(cens, 3, compact_fun = PCAena, compact_args = list(vartot = .7))
#' 
#' # usando diferentes funcoes de clusterizacao
#' selec <- selecporcluster(cens, 3, clust_fun = clusthierarq, method = "single")
#' 
#' # os dois metodos seguintes precisam que os pacotes 'mclust' e 'cluster', respectivamente, 
#' # estejam instalados
#' \dontrun{
#' selec <- selecporcluster(cens, 3, clust_fun = clustEM, modelNames = c("EEE", "VVV"))
#' selec <- selecporcluster(cens, 3, clust_fun = clustkmedoids, metric = "manhattan")
#' }
#' 
#' # Por fim, pode ser feita a visualizacao da escolha
#' selec <- selecporcluster(cens, 3)
#' \dontrun{
#' plot(cens, cens[, , selec[[1]]])
#' }
#' 
#' @seealso \code{\link{plot.cenariosena}} para visualizacao dos cenarios selecionados
#' 
#' @export

selecporcluster <- function(cenarios, nc, clust_fun = clustkmeans, ..., transforma = FALSE,
    compact_fun = PCAena, compact_args = list()) {

    compact_call <- c(list(compact_fun, cenarios), compact_args)
    compact <- eval(as.call(compact_call))

    cluster_call <- c(list(clust_fun, compact, nc), list(...))
    cluster <- eval(as.call(cluster_call))

    if(attr(compact, "teminv") & transforma) {
        stop("Transformacao inversa ainda nao implementada")
    } else {

        # calcula elementos mais proximos de cada centroide

        dat <- extracdims(compact)

        novaordem <- split(seq(nrow(dat)), getclustclass(cluster))

        dat_clust <- split(dat, getclustclass(cluster))
        clstmeans <- getclustmeans(cluster)
        clstmeans <- lapply(seq(dat_clust), function(i) clstmeans[i, ])

        maisprox <- mapply(dat_clust, clstmeans,
            FUN = function(dc, cm) {
                l <- mapply(dc, cm, FUN = "-", SIMPLIFY = FALSE)
                l <- lapply(l, "^", 2)
                Reduce("+", l)
            },
        SIMPLIFY = FALSE)
        maisprox <- lapply(maisprox, which.min)
        maisprox <- mapply(novaordem, maisprox, FUN = function(v, i) v[i])

        out <- list(maisprox, compact, cluster)

        return(out)
    }
}

R/compactacao.r
############################# FUNCOES PARA REDUCAO DE DIMENSIONALIDADE #############################

#' Funcoes de compactacao
#' 
#' Wrappers de metodos de compactacao para uso nos processos desse pacote
#' 
#' Durante a execucao das funcoes de selecao e/ou extracao de cenarios, o objeto \code{cenariosena}
#' passa por uma reducao de dimensionalidade e posterior clusterizacao. As funcoes para compactacao
#' do dado podem implementar qualquer metodo, mas devem respeitar alguns padroes de argumentos de 
#' entrada e caracteristicas na saida.
#' 
#' Quanto as entradas, qualquer funcao implementada com este proposito deve receber na primeira 
#' posicao o argumento \code{cenarios}, um objeto da classe \code{cenariosena}. Outros argumentos
#' especificos a cada metodo de compactacao podem seguir o primeiro, com qualquer nome (estes 
#' serao passados a chamada via o argumento \code{compact_args} das funcoes de selecao/extracao).
#' 
#' Os objetos retornados pelas funcoes de compactacao devem ser da classe \code{compactcen}. A 
#' geracao destes objetos e feita pela funcao \code{\link{new_compactcen}}. Essencialmente deve 
#' receber um datatable com estrutura igual a de objetos \code{cenariosena}, exceto pela coluna 
#' \code{data} que muda para \code{ind}, indicando o indice da variavel compactada para cada 
#' cenario. Consulte a pagina de ajuda do construtor para mais detalhes.
#' 
#' Atualmente o pacote fornece duas opcoes:
#' 
#' \itemize{
#' \item{\code{\link{PCAena}}}
#' \item{\code{\link{acumulaena}}}
#' }
#' 
#' As paginas de help de cada uma das funcoes contem detalhes a respeito dos argumentos de cada uma
#' (que podem ser passados as funcoes de selecao atraves de \code{compact_args} naquelas chamadas).
#' 
#' @seealso Funcoes \code{\link{PCAena}} e \code{\link{acumulaena}} para compactacao de dados
#' 
#' @name compact_funs
NULL

#' Compactacao Por PCA
#' 
#' Reduz dimensionalidade dos cenarios via PCA mantendo um minimo de variacao total
#' 
#' A matriz de dados considerada para reducao de dimensionalidade tem em cada linha o cenario e,
#' em cada coluna, os passos a frente de simulacao por bacia. Desta forma, cem cenarios 15 passos a
#' frente de duas bacias corresponde a uma matriz de dados 100 por 30. Serao selecionadas as n 
#' primeiras componentes principais que representem no minimo \code{vartot} por cento da variacao
#' total. Deve ser observado que o dado e normalizado para media zero e variancia um antes da 
#' compactacao.
#' 
#' @param cenarios objeto da classe \code{cenariosena} contendo apenas um ano de referencia
#' @param vartot percentual em formato decimal de variacao total mininima
#' 
#' @return objeto da classe \code{compactcen} contendo o dado em dimensao reduzida. Este e uma lista
#'     de um elemento nomeado \code{"compact"}, um \code{data.table} com as colunas
#' 
#' \describe{
#' \item{\code{anoref}}{ano de referencia para geracao do cenario}
#' \item{\code{bacia}}{bacia a qual a ENA corresponde}
#' \item{\code{cenario}}{indice do cenario com respeito a bacia e ano de referencia}
#' \item{\code{ind}}{indice do elemento no vetor de dimensao reduzida}
#' \item{\code{ena}}{valor de energia afluente}
#' }
#' 
#' Adicionalmente tem os atributos
#' 
#' \itemize{
#' \item{\code{metodo}: }{string do nome da funcao chamada para compactacao}
#' \item{\code{teminv}: }{booleano indicando se a compactacao possui inversa}
#' \item{\code{invfunc}: }{caso tenha inversa, a funcao que recebe vetores no espaco compactado e 
#'     retorna no espaco original}
#' }
#' 
#' @examples 
#' 
#' # usando o dado exemplo do pacote
#' 
#' # a compactacao deve ser usada para cada ano de referencia individualmente
#' cens <- cenariosdummy["A1"]
#' 
#' # Compactando apenas os cenarios relativos ao SIN completo
#' cens_compact <- PCAena(cens[, "SIN"])
#' \dontrun{
#' plot(cens_compact)
#' }
#' 
#' @seealso \code{\link{plot.compactcen}} para visualizacao das compactacoes
#' 
#' @export

PCAena <- function(cenarios, vartot = .8) {

    ind <- anoref <- bacia <- cenario <- NULL

    if(vartot > 1) vartot <- vartot / 100

    dat <- copy(cenarios$cenarios)

    pca <- dcast(dat, cenario ~ bacia + data, value.var = "ena")[, -1]
    pca <- prcomp(pca, scale = TRUE)

    if(vartot < 0) {
        importance <- 1
    } else {
        importance <- which(summary(pca)$importance[3, ] >= vartot)[1]
    }

    compdat <- pca$x[, seq(importance), drop = FALSE]
    compdat <- cbind(cenario = seq(nrow(compdat)), as.data.table(compdat))
    compdat <- melt(compdat, id.vars = "cenario", variable.name = "ind", value.name = "ena")
    compdat[, ind := as.numeric(sub("[[:alpha:]]*", "", ind))]

    out <- cbind(anoref = dat$anoref[1], bacia = paste0(attr(cenarios, "bacias"), collapse = "."), compdat)
    setorder(out, anoref, bacia, cenario, ind)

    new_compactcen(out, "PCAena", invtransfpca(pca, importance))
}

#' Calcula ENA Acumulada
#' 
#' Reducao de dimensao por ENA acumulada, possivelmente em partes crescentes
#' 
#' Esta funcao reduz a dimensao de um vetor de cenarios de ENA calculando a soma acumulada do vetor,
#' possivelmente em partes. Deve ser notado que, antes da compactacao, os cenarios sao escalonados
#' para o intervalo \eqn{[0, 1]} de modo a obter valores de menor variancia. Esta regularizacao e feita 
#' por bacia individualmente. Isto objetiva determinar um espaco reduzido que capture principalmente
#' o perfil de cada cenario em cada regiao, independentemente da magnitude.
#' 
#' Se \code{quebras = 3L}, por exemplo, a funcao quebra o cenario em tres
#' partes iguais e calcula a soma de cada um. O vetor em dimensao reduzida e, entao, a soma 
#' acumulada destas partes. Para cenarios de um ano, indo de janeiro a dezembro, o uso de 
#' \code{quebras = 3L} representa o calculo da ENA acumulada ate abril, agosto e dezembro (por 
#' construcao um vetor de valores crescentes).
#' 
#' @param cenarios objeto da classe \code{cenariosena} contendo apenas um ano de referencia
#' @param quebras ou um inteiro indicando em quantas partes iguais separar o dado ou um vetor de 
#'     inteiros indicando as posicoes nas quais separar
#' 
#' @return objeto da classe \code{compactcen} contendo o dado em dimensao reduzida. Este e uma lista
#'     de um elemento nomeado \code{"compact"}, um \code{data.table} com as colunas
#' 
#' \describe{
#' \item{\code{anoref}}{ano de referencia para geracao do cenario}
#' \item{\code{bacia}}{bacia a qual a ENA corresponde}
#' \item{\code{cenario}}{indice do cenario com respeito a bacia e ano de referencia}
#' \item{\code{ind}}{indice do elemento no vetor de dimensao reduzida}
#' \item{\code{ena}}{valor de energia afluente}
#' }
#' 
#' Adicionalmente tem os atributos
#' 
#' \itemize{
#' \item{\code{metodo}: }{string do nome da funcao chamada para compactacao}
#' \item{\code{teminv}: }{booleano indicando se a compactacao possui inversa}
#' \item{\code{invfunc}: }{caso tenha inversa, a funcao que recebe vetores no espaco compactado e 
#'     retorna no espaco original}
#' }
#' 
#' @examples 
#' 
#' # usando o dado exemplo do pacote
#' 
#' # a compactacao deve ser usada para cada ano de referencia individualmente
#' cens <- cenariosdummy["A1"]
#' 
#' # Compactando apenas os cenarios relativos ao SIN completo
#' cens_compact <- acumulaena(cens[, "SIN"], quebras = 2L)
#' \dontrun{
#' plot(cens_compact)
#' }
#' 
#' @seealso \code{\link{plot.compactcen}} para visualizacao das compactacoes
#' 
#' @export

acumulaena <- function(cenarios, quebras = 1L) {

    ena <- ena2 <- NULL

    if(quebras < 0) quebras <- length(attr(cenarios, "datas"))

    dat <- copy(cenarios$cenarios)

    # por algum motivo inexplicavel, ena := da erro mas assim funciona
    dat[, ena2 := (ena - min(ena)) / diff(range(ena)), by = "bacia"]
    dat[, ena := ena2]
    dat[, ena2 := NULL]

    out <- dat[, list(ind = seq(quebras), ena = stepcumsum(ena, quebras)),
        by = c("anoref", "bacia", "cenario")]

    new_compactcen(out, "acumulaena", NULL)
}

# HELPERS ------------------------------------------------------------------------------------------

#' Funcao Inversa Do PCA
#' 
#' Retorna uma funcao que recebe vetores no espaco reduzido e projeta de volta no espaco original
#' 
#' @param pca objeto da classe \code{prcomp}
#' @param importance um inteiro indicando a ultima componente principal a manter
#' 
#' @return funcao que recebe um vetor no espaco reduzido e retorna projecao no espaco original

invtransfpca <- function(pca, importance) {
    SIGMA <- pca$rotation[, seq(importance)]
    SIGMA <- t(SIGMA)
    out <- function(newdata) newdata %*% SIGMA
    return(out)
}

#' Soma Acumulada Por Degraus
#' 
#' Calcula a soma acumulada de um vetor em partes
#' 
#' @param x vetor do qual calcular soma
#' @param qbr escalar ou vetor de inteiros indicando quantas partes ou onde separar as partes
#' 
#' @return vetor de ENA acumulada por partes

stepcumsum <- function(x, qbr) {
    if(length(qbr) == 1) qbr <- floor(seq(1, length(x), length.out = qbr + 1)[-1])
    out <- sapply(qbr, function(q) sum(x[seq(q)]))
    out <- unname(out)
    return(out)
}


R/classe-cenariosena.r
########################## FUNCOES PARA IMPORTACAO E MANIPULACAO DOS DADOS #########################

#' Leitura Do Arquivo De Cenarios
#' 
#' Funcao auxiliar que le arquivo texto e formata como objeto da classe \code{cenariosena}
#' 
#' Espera-se um arquivo com formato especifico. As tres primeiras colunas devem conter o indice do 
#' cenario, bacia e ano de referencia, respectivamente. As demais colunas devem corresponder aos
#' passos de tempo para os quais os cenarios foram gerados. Por exemplo, um arquivo contendo cinco 
#' cenarios gerados para 2022-01, 2022-02 e 2022-03 teria a forma
#' 
#' | cenario | bacia | anoref | 1/2022 | 2/2022 | 3/2022 |
#' | :-----: | :---: | :----: | :----: | :----: | :----: |
#' | 1       | SIN   | AA     | XXXX   | XXXX   | XXXX   | 
#' | 2       | SIN   | AA     | XXXX   | XXXX   | XXXX   | 
#' | 3       | SIN   | AA     | XXXX   | XXXX   | XXXX   |
#' | 4       | SIN   | AA     | XXXX   | XXXX   | XXXX   |
#' | 5       | SIN   | AA     | XXXX   | XXXX   | XXXX   |
#' 
#' XXXX   e um placeholder para os valores de ena e AA um placeholder para os valores de ano de 
#' referencia. Nao existe um padrao para os anos, podendo ser qualquer tipo de dado (i.e. "A1", 
#' "ano_1", 2022 e etc).
#' 
#' O nome das colunas de cenarios nao precisa seguir esta formato, sendo possivel informar o padrao
#' de nome pelo argumento \code{pat_data}, que deve corresponder a uma string de formato de data.
#' Neste caso, \code{pat_data = "%m/$Y"}
#' 
#' @param arq caminho do arquivo a ser lido com os cenarios. Formato deve ser compativel com o 
#'     descrito em Detalhes
#' @param pat_data string indicando o formato de data dos nomes das colunas de cenario
#' 
#' @examples 
#' 
#' # leitura do arquvo embutido no pacote
#' arq <- system.file("extdata/cenarios.csv", package = "clustena")
#' cens <- learqcenarios(arq)
#' 
#' @return objeto da classe \code{cenariosena} contendo dados do arquivo lido
#' 
#' @seealso \code{\link{[.cenariosena}} para subset de objetos da classe \code{cenariosena}
#' 
#' @export

learqcenarios <- function(arq, pat_data = "%m/%Y") {

    cenario <- data <- bacia <- anoref <- NULL

    temdia <- grepl("\\%d", pat_data)
    pat_data_2 <- ifelse(temdia, pat_data, paste0("%d__", pat_data))

    dat <- fread(arq)

    coldata <- as.Date(paste0("01__", colnames(dat)), format = pat_data_2)
    colnames(dat)[!is.na(coldata)] <- as.character(coldata[!is.na(coldata)])

    dat[, cenario := seq(.N), by = c("bacia", "anoref")]

    dat <- melt(dat, id.vars = colnames(dat)[is.na(coldata)], value.name = "ena", variable.name = "data")
    dat[, data := as.Date(as.character(data))]
    setorder(dat, bacia, anoref, cenario)

    new_cenariosena(dat)
}

new_cenariosena <- function(dat) {

    out <- list(cenarios = dat[, .SD, .SDcols = c("anoref", "bacia", "cenario", "data", "ena")])

    class(out) <- "cenariosena"

    attr(out, "bacias") <- unique(dat$bacia)
    attr(out, "datas")  <- unique(dat$data)
    attr(out, "anos")   <- unique(dat$anoref)
    attr(out, "ncen")   <- nrow(dat) / with(attributes(out), length(bacias) * length(datas) * length(anos))

    return(out)
}

#' Conversao Para \code{cenariosena}
#' 
#' Converte um data.table em um objeto da classe \code{cenariosena}
#' 
#' \code{dat} deve ter a seguinte estrutura
#' 
#' | anoref | bacia | cenario | indice | valor |
#' | :----: | :---: | :-----: | :----: | :---: |
#' | AA     | SIN   | 1       | 1      | XXXX  | 
#' | AA     | SIN   | 1       | 2      | XXXX  | 
#' | AA     | SIN   | 1       | 3      | XXXX  |
#' | AA     | SIN   | 2       | 1      | XXXX  |
#' | AA     | SIN   | 2       | 2      | XXXX  |
#' 
#' @param dat data.table para converter
#' 
#' @return objeto da classe \code{cenariosena}
#' 
#' @export

as.cenariosena <- function(dat) new_cenariosena(dat)

# METODOS ------------------------------------------------------------------------------------------

#' Subset De \code{cenariosena}
#' 
#' Extrai subsets de objetos \code{cenariosena} e adequa os atributios
#' 
#' @param x objeto da classe \code{cenariosena}
#' @param i escalar ou vetor de anos de referencia para extrair do dado
#' @param j escalar ou vetor de bacias para extrair do dado
#' @param k escalar ou vetor de indices de cenarios para extrair do dado
#' @param l escalar ou vetor de datas para extrair do dado
#' @param ... nao possui uso -- existe apenas para consistencia com a generica
#' 
#' @return objecto \code{cenariosena} contendo apenas os valores especificados
#' 
#' @examples 
#' 
#' # utilizando o dado exemplo do pacote
#' 
#' # subset apenas do ano de referencia "A1"
#' cens <- cenariosdummy["A1"]
#' 
#' # subset apenas da regiao "SUL"
#' cens <- cenariosdummy[, "SUL"]
#' 
#' # subset dos cenarios 5, 10, 20
#' cens <- cenariosdummy[, , c(5, 10, 20)]
#' 
#' # datas simuladas 2022-11-01/2023-04-01
#' cens <- cenariosdummy[, , , seq.Date(as.Date("2022-11-01"), as.Date("2023-04-01"), by = "month")]
#' 
#' # ano "A2", NE e N, cenarios 10, 11, 12 nas datas 2022-09-01/2023-02-01
#' cens <- cenariosdummy[
#'     "A2",
#'     c("NE", "N"),
#'     c(10, 11, 12),
#'     seq.Date(as.Date("2022-09-01"), as.Date("2023-02-01"), by = "month")
#' ]
#' 
#' @export

`[.cenariosena` <- function(x, i, j, k, l, ...) {

    anoref <- bacia <- cenario <- data <- NULL

    dat <- copy(x$cenarios)

    if(missing(i)) {
        i <- attr(x, "anos")
    }
    if(missing(j)) {
        j <- attr(x, "bacias")
    } else if(is.numeric(j)) {
        j <- attr(x, "bacias")[j]
    }
    if(missing(k)) {
        k <- seq(attr(x, "ncen"))
    }
    if(missing(l)) {
        l <- attr(x, "datas")
    } else if(is.numeric(l)) {
        l <- attr(x, "datas")[l]
    } else {
        l <- as.Date(l)
    }

    dat <- dat[(anoref %in% i) & (bacia %in% j) & (cenario %in% k) & (data %in% l)]

    new_cenariosena(dat)
}
